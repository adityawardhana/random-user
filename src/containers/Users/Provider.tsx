import React, { FC, useEffect, useState } from "react";
import qs from "query-string";
import { getUsers } from "../../services/users";
import useFetch from "../../hooks/useFetch";
import useDebounce from "../../hooks/useDebounce";
import UserContext from "./Context";

const Provider: FC<{}> = ({ children }) => {
  const [params, setParams] = useState({
    inc: "gender,name,login,email,registered",
    results: 10,
    page: 1,
    keyword: "",
    sortBy: "",
    sortOrder: "",
    gender: "",
  });
  const [keyword, setKeyword] = useState("");
  const [searchFocus, setSearchFocus] = useState(false);

  const {
    fetch,
    loading,
    error,
    data: { results: users } = {},
  } = useFetch(
    (newParams) =>
      getUsers(qs.stringify(newParams || params, { skipEmptyString: true })),
    { immediate: false }
  );

  const handleChangeSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
    setKeyword(e.target.value);
  };

  const handleSelectGender = (value: string) => {
    setParams({ ...params, gender: value });
  };

  const handleChangeSort = (sortBy: string, sortOrder: string) => {
    setParams({ ...params, sortBy, sortOrder, page: 1 });
  };

  const handleChangePage = (page: number) => {
    setParams({ ...params, page });
  };

  const handleResetFilter = () => {
    setKeyword("");
    setParams({ ...params, page: 1, keyword: "", gender: "" });
  };

  useDebounce(
    () => {
      searchFocus && setParams({ ...params, keyword, page: 1 });
    },
    keyword,
    1000
  );

  useEffect(() => {
    fetch({ ...params });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [params]);

  return (
    <UserContext.Provider
      value={{
        params,
        users,
        loading,
        error,
        keyword,
        setSearchFocus,
        handleChangePage,
        handleChangeSearch,
        handleChangeSort,
        handleSelectGender,
        handleResetFilter,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export default Provider;
