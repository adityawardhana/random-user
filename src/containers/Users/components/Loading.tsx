import { Box, Spinner } from "@adityawardhana/ts-react-ui";
import useUsers from "../useUser";

const Loading = () => {
  const { loading } = useUsers();
  if (!loading) {
    return null;
  }
  return (
    <Box height={400} alignItems="center" justifyContent="center">
      <Box width={50} height={50} m="auto">
        <Spinner color="secondary" />
      </Box>
    </Box>
  );
};

export default Loading;
