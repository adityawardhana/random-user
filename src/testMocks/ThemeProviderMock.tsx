import React from "react";
import theme, { GlobalStyles } from "@adityawardhana/ts-react-ui/lib/theme";
import { ThemeProvider } from "styled-components";

export const setupTheme = (children: React.ReactNode) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};
