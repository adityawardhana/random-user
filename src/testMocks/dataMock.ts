export const resultAPIMock = {
  login: {
    username: "adityawrdhn",
  },
  name: {
    first: "adit",
    last: "w",
  },
  email: "adityawrdhn@gmail.com",
  gender: "male",
  registered: { date: "2003-08-26T21:56:32.575Z" },
};

export const usersMock = {
  params: { sortBy: "email", sortOrder: "ascend", page: 1, gender: "" },
  loading: false,
  users: [...Array(3)].map(() => ({ ...resultAPIMock })),
  keyword: "xxx",
  handleChangeSort: jest.fn(),
  handleChangePage: jest.fn(),
  handleChangeSearch: jest.fn(),
  setSearchFocus: jest.fn(),
  handleSelectGender: jest.fn(),
  handleResetFilter: jest.fn(),
};
