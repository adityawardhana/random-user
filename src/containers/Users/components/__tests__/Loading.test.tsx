import { render } from "@testing-library/react";
import { usersMock } from "../../../../testMocks/dataMock";
import { setupTheme } from "../../../../testMocks/ThemeProviderMock";
import Provider from "../../Provider";
import useUsersMock from "../../useUser";
import Loading from "../Loading";

jest.mock("../../useUser", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("Loading unit tests", () => {
  test("should not render loading", () => {
    (useUsersMock as any).mockReturnValue({ ...usersMock });

    const { container } = render(
      setupTheme(
        <Provider>
          <Loading />
        </Provider>
      )
    );
    expect(container).toMatchSnapshot();
  });

  test("should render loading", () => {
    (useUsersMock as any).mockReturnValue({ ...usersMock, loading: true });

    const { container } = render(
      setupTheme(
        <Provider>
          <Loading />
        </Provider>
      )
    );
    expect(container).toMatchSnapshot();
  });
});
