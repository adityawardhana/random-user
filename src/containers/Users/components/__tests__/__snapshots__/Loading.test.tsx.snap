// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading unit tests should not render loading 1`] = `<div />`;

exports[`Loading unit tests should render loading 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 400px;
}

.c1 {
  width: 50px;
  height: 50px;
  margin: auto;
}

.c2 {
  position: relative;
  height: 100%;
  -webkit-animation: kZndMJ 2s infinite linear;
  animation: kZndMJ 2s infinite linear;
}

.c2 circle {
  -webkit-animation: 1.4s ease-in-out infinite both dbIXyf;
  animation: 1.4s ease-in-out infinite both dbIXyf;
  display: block;
  fill: transparent;
  stroke: #152E3B;
  stroke-linecap: round;
  stroke-dasharray: 283;
  stroke-dashoffset: 280;
  stroke-width: 10px;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
}

<div>
  <div
    class="c0"
    height="400"
  >
    <div
      class="c1"
      height="50"
      width="50"
    >
      <svg
        class="c2"
        color="secondary"
        viewBox="0 0 100 100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="50"
          cy="50"
          r="45"
        />
      </svg>
    </div>
  </div>
</div>
`;
